Document/0x06g-Testing-Network-Communication.md

```diff
@@ -275,11 +277,42 @@ Sometimes applications have one certificate that is hardcoded and use it for the
 
 Once you have extracted the certificate from the application (e.g. using Cycript or Frida), add it as client certificate in Burp, and you will be able to intercept the traffic.
 
+##### Bypassing Certificate Pinning
+
+There are various ways to bypass SSL Pinning and the following section will describe it for jailbroken and non-jailbroken devices.
+
+If you have a jailbroken device you can try the following:
+
+- "[SSL Kill Switch 2](https://github.com/nabla-c0d3/ssl-kill-switch2 "SSL Kill Switch 2")" is one way to disable certificate pinning. It can be installed via the Cydia store. It will hook on to all high-level API calls and bypass certificate pinning.
+- The Burp Suite app "[Mobile Assistant](https://portswigger.net/burp/help/mobile_testing_using_mobile_assistant.html "Using Burp Suite Mobile Assistant")" can also be used to bypass certificate pinning.
+
+In some cases, certificate pinning is tricky to bypass. Look for the following when you can access the source code and recompile the app:
+
+- the API calls `NSURLSession`, `CFStream`, and `AFNetworking`
+- methods/strings containing words like "pinning," "X.509," "Certificate," etc.
+
+If you don't have access to the source, you can try binary patching:
+
+- If OpenSSL certificate pinning is used, you can try [binary patching](https://www.nccgroup.trust/us/about-us/newsroom-and-events/blog/2015/january/bypassing-openssl-certificate-pinning-in-ios-apps/ "Bypassing OpenSSL Certificate Pinning in iOS Apps").
+- Sometimes, the certificate is a file in the application bundle. Replacing the certificate with Burp's certificate may be sufficient, but beware of the certificate's SHA sum. If it's hardcoded into the binary, you must replace it too!
+
+It is also possible to bypass SSL Pinning on non-jailbroken devices by using Frida and objection (this also works on jailbroken devices). As a prerequisite the iOS app would need to be repackaged and signed, which can be automated through objection (please take note that this can only be done on macOS with Xcode). For detailed information please visit the objection GitHub Wiki on [how to repackage](https://github.com/sensepost/objection/wiki/Patching-iOS-Applications "Patching iOS Applications"). By using the following command in objection you can disable common SSL Pinning implementations:
+
+```shell
+$ ios sslpinning disable
+```
+
+You can look into the [pinning.ts](https://github.com/sensepost/objection/blob/master/agent/src/ios/pinning.ts "pinning.ts") file to understand how the bypass works.
+
+See also the [GitHub Page](https://github.com/sensepost/objection#ssl-pinning-bypass-running-for-an-ios-application "Disable SSL Pinning in iOS" ) for further information. 
+
+If you want to get more details about white box testing and typical code patterns, refer to "iOS Application Security" by David Thiel. It contains descriptions and code snippets illustrating the most common certificate pinning techniques.
```

Comment:
so where do we get that book? are the references moved properly as well? or is there another way to get it?

Reply:
added the book as reference according to the style guide 
https://github.com/OWASP/owasp-mstg/blob/master/style_guide.md#3-external-references